Add autocut and relative score ranking algo to hybrid search

dirkkul opened the pull request:
### What's being changed:

// FusionRelativeScore uses the relative differences in the scores from keyword and vector search to combine the
// results. This method retains more information than ranked fusion and should result in better results.
//
// The scores from each result are normalized between 0 and 1, e.g. the maximum score becomes 1 and the minimum 0 and the
// other scores are in between, keeping their relative distance to the other scores.
// Example:
//
//	Input score = [1, 8, 6, 11] => [0, 0.7, 0.5, 1]
//
// The normalized scores are then combined using their respective weight and the combined scores are sorted

These scores can then be used for autocut.

Usage:
```
{
  Get {
    CLASS(
      hybrid: {autocut: 2, query: "nice day", alpha: 0.5, vector: [4, 5, 4], fusion_type: relative_score_fusion}
    ) {
      prop
      _additional {
        score
        vector
      }
    }
  }
}

```
- autocut to enabled autocut
- fusion_type: relative_score_fusion to enable new fusion algo, default is ranked_fusion
Can be used independently from each other

### Review checklist
- [x] All new code is covered by tests where it is reasonable.

aliszka commented:
```
BM25 scores: {1:50, 3:25, 4:1, 2:0} => {1:1, 3:0.5, 4:0.02, 2:0.002}
Vector scores: {4:3, 2:2, 1:0.6, 3:0} => {4:1, 2:0.666, 1:0.2, 3:0.1}
```
I don't quite understand why zeros become positive numbers after normalization in this example (2:0 -> 2:0.002, 3:0 -> 3:0.1), yet when provided to `FusionRelativeScore` method they stay as `0`s.

What would be expected result of normalization {10, 5, 5}? I would expect {1, 0.5, 0.5}, but according to FusionRelativeScore it is {1, 0, 0}

dirkkul commented on usecases/traverser/hybrid/hybrid_fusion.go:
```diff
@@ -102,19 +70,53 @@ func FusionReciprocal(weights []float64, results [][]*Result) []*Result {
-		concatenatedResults = append(concatenatedResults, result...)
+func FusionRelativeScore(weights []float64, results [][]*Result) []*Result {
+	maximum := []float32{-100000, -100000}
+	minimum := []float32{100000, 100000}
```
1) good point, changed
2) If alpha is 0 or 1 it is just one entry. Also added a test for this case

dirkkul commented on usecases/traverser/hybrid/hybrid_fusion.go:
```diff
@@ -102,19 +70,53 @@ func FusionReciprocal(weights []float64, results [][]*Result) []*Result {
+		weight := float32(weights[i])
+		for _, res := range results[i] {
+			tempResult := res
+			score := weight * (res.SecondarySortValue - minimum[i]) / (maximum[i] - minimum[i])
```
fixed

dirkkul commented on usecases/traverser/hybrid/hybrid_fusion.go:
```diff
@@ -102,19 +70,53 @@ func FusionReciprocal(weights []float64, results [][]*Result) []*Result {
+	for i := range results {
+		weight := float32(weights[i])
+		for _, res := range results[i] {
+			tempResult := res
```
not needed, removed it

dirkkul commented:
The text was a bit older and I changed the implementation a bit since then. {10, 5, 5} becomes {1, 0, 0}, I updated the PR description and added more comments to the code. Let me know if anything else is unclear

dirkkul commented on usecases/traverser/hybrid/hybrid_fusion.go:
```diff
@@ -0,0 +1,143 @@
+//
+// The normalized scores are then combined using their respective weight and the combined scores are sorted
+func FusionRelativeScore(weights []float64, results [][]*Result) []*Result {
+	if len(results[0]) == 0 {
```
thinking again, it doesn't. Added tests and fixed it :)

dirkkul commented on usecases/traverser/hybrid/hybrid_fusion.go:
```diff
@@ -0,0 +1,143 @@
+//	Input score = [1, 8, 6, 11] => [0, 0.7, 0.5, 1]
+//
+// The normalized scores are then combined using their respective weight and the combined scores are sorted
+func FusionRelativeScore(weights []float64, results [][]*Result) []*Result {
```
added it in the caller, so the other fusion algo also profits from the check

aliszka commented on usecases/traverser/hybrid/hybrid_fusion.go:
```diff
@@ -0,0 +1,145 @@
+//
+// The normalized scores are then combined using their respective weight and the combined scores are sorted
+func FusionRelativeScore(weights []float64, results [][]*Result) []*Result {
+	if len(results[0]) == 0 && len(results[1]) == 0 {
```
it is bound to results size. will fail for array with single element of 0 size.

aliszka commented on usecases/traverser/hybrid/hybrid_fusion.go:
```diff
@@ -0,0 +1,145 @@
+		return []*Result{}
+	}
+
+	var maximum []float32
+	var minimum []float32
+
+	for i := range results {
+		if len(results[i]) > 0 {
+			maximum = append(maximum, results[i][0].SecondarySortValue)
+			minimum = append(minimum, results[i][0].SecondarySortValue)
+		} else { // dummy values so the indices match
+			maximum = append(maximum, 0)
+			minimum = append(minimum, 0)
+		}
+		for _, res := range results[i] {
+			if res.SecondarySortValue > maximum[i] {
+				maximum[i] = res.SecondarySortValue
+			}
+
+			if res.SecondarySortValue < minimum[i] {
+				minimum[i] = res.SecondarySortValue
+			}
+		}
+	}
```
🔝 would work, below IMO simpler version without dummy appending
```
// already inited by 0s
maximum := make([]float32, len(results))
minimum := make([]float32, len(results))

for i := range results {
	for j, res := range results[i] {
		if j == 0 || res.SecondarySortValue > maximum[i] {
			maximum[i] = res.SecondarySortValue
		}
		if j == 0 || res.SecondarySortValue < minimum[i] {
			minimum[i] = res.SecondarySortValue
		}
	}
}
```

aliszka commented on usecases/traverser/hybrid/hybrid_fusion.go:
```diff
@@ -0,0 +1,145 @@
+	for i := range results {
+		weight := float32(weights[i])
+		for _, res := range results[i] {
+			var score float32
+			if maximum[i] != minimum[i] {
+				score = weight * (res.SecondarySortValue - minimum[i]) / (maximum[i] - minimum[i])
+			} else {
+				score = 0
+			}
```
`var score float32` is already 0, but to be more explicit it can be set as `score := 0` and `else {...}` could be removed
